### ISSUE TITLE
Error during WebSocket handshake: Unexpected response code: 404

### ISSUE THREAD

**Issue by papaytl185** [aspnet/SignalR#984]
I am running my mvc aspnetcore app in a server farm with Redis and am having similar issues with a 404 error. I'm not sure where to look, we are running pfsense as the load balancer with iis servers. Its pretty consistent that every other refresh successfully connects and the other times I get the 404. If I take 1 server offline, leaving just one, it always works. And vice versa if i take the other server offline it works, so both servers work fine when running through the load balancer 1 at a time.

Any help would be greatly appreciated as I am not sure where to look.

WebSocket connection to 'wss://mywebsite.com/notification?id=d0fc384a-14f0-4436-aaaf-e5d50502399f' failed: Error during WebSocket handshake: Unexpected response code: 404

**Comment by davidfowl** [comment]
Sticky sessions are required. Seems like pfsense has docs here https://doc.pfsense.org/index.php/Inbound_Load_Balancing#Sticky_connections

**Comment by papaytl185** [comment]
I wonder why it cant complete the handshake on the other server the request is coming into. Is it due to needing to create a new connectionId or something.. It doesn't seem like a good design forcing sticky sessions on a server farm. With the old signalr i have no issues and don't need sticky sessions. This seems like a step backwards, or I just don't fully understand why I cant make a new handshake. Can I close the old connection and allow the handshake to happen?

Also discussing in this thread.
https://github.com/aspnet/SignalR/issues/868

I turned sticky sessions on and am still getting the issue. hmm

**Comment by yosigolan** [comment]
Hey.
i am working with a single server and receives the same error. after few connection retries it connects successfully. on the server side i dont see any error message.
in order to deal with it for now i have implemented a connection retry mechanism.

**Comment by moozzyk** [comment]
@papaytl185 - it would be good to see logs from the server.

**Comment by papaytl185** [comment]
I finally got it working! It was tough to figure out, but long story short, Sticky sessions was required like @davidfowl said, however, at my company we have a load balancer on our internet traffic as well. So one request was going to one server, while the other was going to the other therefore throwing the 404. I setup my machine to only use one internet connection and it worked like a charm.

Now i need to use a different type of load balancer that does the balancing by cookie or some identifier instead of ip address which is what pfsense uses.

**Comment by yosigolan** [comment]
just to clarify, i am not getting this error all the time, only once in a while.
there is nothing in the server logs in my case.

**Comment by mcliment** [comment]
I had intermittent 404 errors connecting to the hubs using websockets on Azure and activating ARR Affinity solved the problem.

**Comment by moozzyk** [comment]
As a side note - I merged a [change](caff492cdc8eccc42254ec7d584ab48b1fb019fe) (post-alpha2) where if you explicitly chose the WebSocket transport the initial OPTIONS request will not be sent and therefore stickiness won't be required.

**Comment by steco** [comment]
I came across exactly this too. If the transport is not specified, the initial OPTIONS request will be returned from the loadbalancer with a Set-Cookie header - this needs to be carried across to the subsequent GET request otherwise it will get routed to a different host.  HttpConnection/WebSocketsTransport don't do this - are there any plans to fix this ?

I have a hacked-together example which achieves this, but I had to make quite a few assumptions to get it to work.  I am not familiar enough with the codebase to know how to integrate it properly, but happy to share if that would help.  In the meantime I'll use moozzyk's helpful fix and specify the WebSocket transport explicitly.

**Comment by davidfowl** [comment]
> I came across exactly this too. If the transport is not specified, the initial OPTIONS request will be returned from the loadbalancer with a Set-Cookie header - this needs to be carried across to the subsequent GET request otherwise it will get routed to a different host. HttpConnection/WebSocketsTransport don't do this - are there any plans to fix this ?

Are you talking about the .NET client?

**Comment by steco** [comment]
Yes - this is the .NET client against a SignalR hub running in AWS.

**Comment by moozzyk** [comment]
Likely we need to transfer cookies from the OPTIONS request to the WebSocket we are opening for the websockets transport.

**Comment by BrennanConroy** [comment]
Did you double-check if SSE uses the same HttpClient?

**Comment by moozzyk** [comment]
Yes, SSE is using the same HttpClient as the one to send OPTIONS request.

**Comment by davidfowl** [comment]
We need to make a cookie container. That's why it doesn't flow today. We also need to look at the nodejs client

**Comment by steco** [comment]
The main difficulty I had in splicing this into the existing code is that the ctor of HttpConnection needs to pass a CookieContainer into its instance of HttpClient.  This is done using HttpClientHandler, but the ctor of HttpConnection already takes a parameter of HttpMessageHandler (base class of HttpClientHandler) which it uses if specified.  I don't know enough about the other usages of this to know how best to change this.

Once this is done it's simple to extract the cookies from the Negotiate response and pass them to the WebSocketsTransport.

**Comment by steco** [comment]
The above worked for a .NET Core project, but I am having difficulty splicing the same into a .Net Framework project.  Do you have a planned date for the next release (which will include this workaround which I'll be able to use - https://github.com/aspnet/SignalR/issues/1035) ?

**Comment by talarari** [comment]
#1035 is really important, any chance to get a release including it? 
Websockets are practically supported everywhere at this point, requiring sticky sessions really complicates things.

**Comment by davidfowl** [comment]
@talarari yes we will at some point, if you can't wait, the ci builds are always available for use (none of this is production ready anyways). Sticky sessions will still be required for non websockets transports of course and you also need to explicitly declare that you want to use websockets on the client side to avoid the initial request.

**Comment by adam3039** [comment]
If anyone runs into this issue with Azure Service Fabric, or more specifically an Azure load balancer, I fixed it by going into the load balancer rules, selecting the rule for the port SignalR is using (in my case 443), and setting the "Session Persistence" to "Client IP". This can also be done in your ARM template for full service fabric deployments.

**Comment by anurse** [comment]
It looks to me like this issue has been resolved, so I'm going to close it. We've got a one-request WebSocket connection coming in the next release so you can avoid having to enable sticky sessions if you're only supporting WebSockets. For other transports, we require sticky sessions by design as they remove a huge amount of complexity (and a massive source of bugs) in scale-out.

Regarding cookies, we can use #1210 to track adding a better configuration API for that.

Feel free to continue commenting or reopen this if there's something we missed, or file new issues.

**Comment by sinaislam** [comment]
WebSocket connection to 'ws://localhost:15536/message?id=FPgRBnY6DGnZdSMinT5QiQ' failed: Error during WebSocket handshake: Unexpected response code: 404

signalr.js:1883 Error: Failed to start the transport 'WebSockets': undefined


Error: Failed to start the connection: Error: Unable to initialize any of the available transports.


signalr.js:3057 Uncaught (in promise) Error: Unable to initialize any of the available transports.
    at HttpConnection.<anonymous> (signalr.js:3057)
    at step (signalr.js:118)
    at Object.throw (signalr.js:99)
    at rejected (signalr.js:90)

**Comment by anurse** [comment]
This issue is closed. If you need assistance, you should file a new issue. Make sure to follow the [Diagnostics Guide](https://github.com/aspnet/SignalR/wiki/Diagnostics-Guide) and gather client and server side traces.
